controller/state.go

```diff
@@ -137,7 +137,7 @@ func (s *ksonnetAppStateManager) getTargetObjs(app *v1alpha1.Application, revisi
 		}
 		targetObjs = append(targetObjs, obj)
 	}
-	return targetObjs, manifestInfo, nil
```

Comment:
Right now all sync operations are executed in parallel:

https://github.com/argoproj/argo-cd/blob/master/controller/sync.go#L355

So even if objects are sorted then changes will be done in parallel and we cannot guarantee the creation order. I think we should apply creation order in `doApplySync` method. 

I also think we need to delete objects in order as well, so I suggest to make the following changes in `doApplySync` method:

- Split `syncTasks` array into two groups: delete tasks, apply tasks. Groups can be processed in parallel. Tasks of each group should be executed in order based on object kind. 
- we should use [install](https://github.com/helm/helm/blob/release-2.10/pkg/tiller/kind_sorter.go#L29) order for creation and [uninstall](https://github.com/helm/helm/blob/release-2.10/pkg/tiller/kind_sorter.go#L61) for deletion.
- run creation/deletion tasks with same order in parallel. e.g. pods should be deleted in parallel

Reply:
I left the sorting here since the getTargetObjs and getLiveObjs functions are used as the source of truth to get k8s resources.  This information is passed to the doApplySync and allows us to show the resources in the correct order in the UI.  I changed the `doApplySync` method to apply resources one kind at a time.